synapse/handlers/federation.py

```diff
@@ -278,9 +278,15 @@ def on_receive_pdu(self, origin, pdu, sent_to_us_directly=False):
                             len(missing_prevs),
                         )
 
-                        yield self._get_missing_events_for_pdu(
-                            origin, pdu, prevs, min_depth
-                        )
+                        try:
+                            yield self._get_missing_events_for_pdu(
+                                origin, pdu, prevs, min_depth
+                            )
+                        except Exception as e:
+                            raise Exception(
+                                "Error fetching missing prev_events for %s: %s"
+                                % (event_id, e)
+                            )
```

Comment:
I'd be tempted to not include `e` and rely on the exception chaining (`raise Exception(..) from e`).

(I assume we've checked that `get_missing_events_for_pdu` doesn't raise any expected exceptions.)

Reply:
> I'd be tempted to not include e and rely on the exception chaining (raise Exception(..) from e).

I tried that. It made for remarkably unreadable errors, not least because the exception chain is epic.

Exception chaining happens by default when you raise an exception from an `except` block, afaik

Reply:
> (I assume we've checked that get_missing_events_for_pdu doesn't raise any expected exceptions.)

well: first, I've looked at the places `on_receive_pdu` is called (ie, the places that might care that we turn one sort of exception into a generic `Exception`). There are two such places: `_handle_queued_pdus`, in which all exceptions are treated equally, and `get_missing_events_for_pdu` (again), which does treat `FederationError` with `code=403` specially.

Then going *down* through `get_missing_events_for_pdu`, the only things which might raise a `FederationError` are `federation_client.get_missing_events` (which I'm 99% sure doesn't), and `on_receive_pdu` which, as discussed, filters out such exceptions. So I'm reasonably confident it's not possible for get_missing_events_for_pdu to raise a `FederationError` with `code=403`, and therefore this change shouldn't make any difference.